Dataframe constructor misinterprets `columns` argument if nested list is passed in as the `data` parameter.

This issue is based on Stack Overflow question http://stackoverflow.com/q/40182072/2988730.
#### A small, complete example of the issue

``` python
df = pd.DataFrame([[1, 2, 3], [4, 5, 6]],
                  index=[['gibberish']*2, [0, 1]],
                  columns=[['baldersash']*3, [10, 20, 30]])
```

The result is

```
  File "<ipython-input-321-2695882ac68b>", line 3, in <module>
    columns=[['baldersash']*3, [10, 20, 30]])

  File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/pandas/core/frame.py", line 263, in __init__
    arrays, columns = _to_arrays(data, columns, dtype=dtype)

  File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/pandas/core/frame.py", line 5352, in _to_arrays
    dtype=dtype)

  File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/pandas/core/frame.py", line 5431, in _list_to_arrays
    coerce_float=coerce_float)

  File "/home/jfoxrabi/miniconda3/lib/python3.5/site-packages/pandas/core/frame.py", line 5489, in _convert_object_array
    'columns' % (len(columns), len(content)))

AssertionError: 2 columns passed, passed data had 3 columns
```
#### Expected Output

```
            baldersash   
                    10 20 30
gibberish 0          1  2  3
          1          4  5  6
```

The surprising thing here is that any of the following seem to work just fine:
1. Supplying a numpy array as `data`:
   
   ``` python
   df = pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]),
                     index=[['gibberish']*2, [0, 1]],
                     columns=[['baldersash']*3, [10, 20, 30]])
   ```
   
   Results in
   
   ```
               baldersash      
                       10 20 30
   gibberish 0          1  2  3
             1          4  5  6
   ```
2. Reducing the size of the input array to have two columns:
   
   ``` python
   df = pd.DataFrame([[1, 2], [3, 4]],
                     index=[['gibberish']*2, [0, 1]],
                     columns=[['baldersash']*2, [10, 20]])
   ```
   
   Results in
   
   ```
               baldersash   
                       10 20
   gibberish 0          1  2
             1          3  4
   ```
3. Omitting the `columns` argument:
   
   ``` python
   df = pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]),
                     index=[['gibberish']*2, [0, 1]])
   ```
   
   Results in
   
   ```
                0  1  2
   gibberish 0  1  2  3
             1  4  5  6
   ```
4. Using a single-level list for the `columns` argument:
   
   ``` python
   df = pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]),
                     index=[['gibberish']*2, [0, 1]],
                     columns=[10, 20, 30])
   ```
   
   Results in
   
   ```
                10  20  30
   gibberish 0   1   2   3
             1   4   5   6
   ```
#### Output of `pd.show_versions()`

<details>
## INSTALLED VERSIONS

commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-431.29.2.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 25.1.6
Cython: 0.24.1
numpy: 1.11.1
scipy: 0.18.0
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.4.1
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: None
tables: None
numexpr: 2.6.1
matplotlib: 1.5.1
openpyxl: 2.3.5
xlrd: None
xlwt: None
xlsxwriter: 0.8.4
lxml: 3.5.0
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None
</details>

@jreback I didn't know you could pass a list of lists to represent a MultiIndex (like you would pass to `MultiIndex.from_arrays`, but in any case it works in certain cases and it not working in the certain example of @madphysicist is inconsistent:

- So passing a nested list or corresponding array as the *data* gives the same result

```
In [78]: pd.DataFrame([[1, 2, 3], [4, 5, 6]])
Out[78]: 
   0  1  2
0  1  2  3
1  4  5  6

In [79]: pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]))
Out[79]: 
   0  1  2
0  1  2  3
1  4  5  6
```

* The same as above, but passing a nested list to `index=` -> nested list is interpreted as a MultiIndex:
```
In [82]: pd.DataFrame([[1, 2, 3], [4, 5, 6]], index=[['A', 'A'], ['a', 'b']])
Out[82]: 
     0  1  2
A a  1  2  3
  b  4  5  6

In [83]: pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]), index=[['A', 'A'], ['a', 'b']])
Out[83]: 
     0  1  2
A a  1  2  3
  b  4  5  6

In [84]: pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]), index=[['A', 'A'], ['a', 'b']]).index
Out[84]: 
MultiIndex(levels=[['A'], ['a', 'b']],
           labels=[[0, 0], [0, 1]])

```

* The same for `columns=` -> but now there is a difference between having a list vs array for the *data*:
```
In [80]: pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=[['A', 'A', 'A'], ['a', 'b', 'c']])
...
AssertionError: 2 columns passed, passed data had 3 columns

In [81]: pd.DataFrame(np.array([[1, 2, 3], [4, 5, 6]]), columns=[['A', 'A', 'A'], ['a', 'b', 'c']])
Out[81]: 
   A      
   a  b  c
0  1  2  3
1  4  5  6
```

--- a/pandas/tests/frame/methods/test_diff.py
+++ b/pandas/tests/frame/methods/test_diff.py
@@ -286,10 +286,11 @@ class TestDataFrameDiff:
         expected = DataFrame(np.array(df)).diff()
         tm.assert_frame_equal(result, expected)
 
-    def test_diff_all_int_dtype(self, any_int_or_nullable_int_dtype):
+    def test_diff_all_int_dtype(self, any_int_dtype):
         # GH 14773
         df = DataFrame(range(5))
-        df = df.astype(np.int8)
+        df = df.astype(any_int_dtype)
         result = df.diff()
-        expected = DataFrame([np.nan, 1.0, 1.0, 1.0, 1.0], dtype="float32")
+        expected_dtype = "float32" if any_int_dtype in ("int8", "int16") else "float64"
+        expected = DataFrame([np.nan, 1.0, 1.0, 1.0, 1.0], dtype=expected_dtype)
         tm.assert_frame_equal(result, expected)

--- a/pandas/tests/frame/methods/test_diff.py
+++ b/pandas/tests/frame/methods/test_diff.py
@@ -285,3 +285,11 @@ class TestDataFrameDiff:
         result = df.diff()
         expected = DataFrame(np.array(df)).diff()
         tm.assert_frame_equal(result, expected)
+
+    def test_diff_all_int_dtype(self, any_int_or_nullable_int_dtype):
+        # GH 14773
+        df = DataFrame(range(5))
+        df = df.astype(np.int8)
+        result = df.diff()
+        expected = DataFrame([np.nan, 1.0, 1.0, 1.0, 1.0], dtype="float32")
+        tm.assert_frame_equal(result, expected)

--- a/pandas/tests/frame/methods/test_quantile.py
+++ b/pandas/tests/frame/methods/test_quantile.py
@@ -705,5 +705,7 @@ class TestQuantileExtensionDtype:
             }
         )
         result = df[["a", "c"]].quantile(0.5, axis=axis)
-        expected = Series(expected_data, name=0.5, index=Index(expected_index))
+        expected = Series(
+            expected_data, name=0.5, index=Index(expected_index), dtype=np.float64
+        )
         tm.assert_series_equal(result, expected)

--- a/pandas/tests/frame/methods/test_quantile.py
+++ b/pandas/tests/frame/methods/test_quantile.py
@@ -4,6 +4,7 @@ import pytest
 import pandas as pd
 from pandas import (
     DataFrame,
+    Index,
     Series,
     Timestamp,
 )
@@ -650,3 +651,59 @@ class TestQuantileExtensionDtype:
             assert result == expected
         else:
             tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "dtype, expected_data, expected_index, axis",
+        [
+            ["float64", [], [], 1],
+            ["int64", [], [], 1],
+            ["float64", [np.nan, np.nan], ["a", "b"], 0],
+            ["int64", [np.nan, np.nan], ["a", "b"], 0],
+        ],
+    )
+    def test_empty_numeric(self, dtype, expected_data, expected_index, axis):
+        # GH 14564
+        df = DataFrame(columns=["a", "b"], dtype=dtype)
+        result = df.quantile(0.5, axis=axis)
+        expected = Series(
+            expected_data, name=0.5, index=Index(expected_index), dtype="float64"
+        )
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "dtype, expected_data, expected_index, axis, expected_dtype",
+        [
+            ["datetime64[ns]", [], [], 1, "float64"],
+            ["datetime64[ns]", [pd.NaT, pd.NaT], ["a", "b"], 0, "datetime64[ns]"],
+        ],
+    )
+    def test_empty_datelike(
+        self, dtype, expected_data, expected_index, axis, expected_dtype
+    ):
+        # GH 14564
+        df = DataFrame(columns=["a", "b"], dtype=dtype)
+        result = df.quantile(0.5, axis=axis, numeric_only=False)
+        expected = Series(
+            expected_data, name=0.5, index=Index(expected_index), dtype=expected_dtype
+        )
+        tm.assert_series_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "expected_data, expected_index, axis",
+        [
+            [[np.nan, np.nan], range(2), 1],
+            [[], [], 0],
+        ],
+    )
+    def test_datelike_numeric_only(self, expected_data, expected_index, axis):
+        # GH 14564
+        df = DataFrame(
+            {
+                "a": pd.to_datetime(["2010", "2011"]),
+                "b": [0, 5],
+                "c": pd.to_datetime(["2011", "2012"]),
+            }
+        )
+        result = df[["a", "c"]].quantile(0.5, axis=axis)
+        expected = Series(expected_data, name=0.5, index=Index(expected_index))
+        tm.assert_series_equal(result, expected)

--- a/pandas/tests/frame/methods/test_quantile.py
+++ b/pandas/tests/frame/methods/test_quantile.py
@@ -673,7 +673,14 @@ class TestQuantileExtensionDtype:
     @pytest.mark.parametrize(
         "dtype, expected_data, expected_index, axis, expected_dtype",
         [
-            ["datetime64[ns]", [], [], 1, "float64"],
+            pytest.param(
+                "datetime64[ns]",
+                [],
+                [],
+                1,
+                "datetime64[ns]",
+                marks=pytest.mark.xfail(reason="#GH 41544"),
+            ),
             ["datetime64[ns]", [pd.NaT, pd.NaT], ["a", "b"], 0, "datetime64[ns]"],
         ],
     )

--- a/pandas/tests/frame/test_constructors.py
+++ b/pandas/tests/frame/test_constructors.py
@@ -2681,3 +2681,14 @@ class TestFromScalar:
         result = constructor(scalar)
 
         assert type(get1(result)) is cls
+
+    def test_nested_list_columns(self):
+        # GH 14467
+        result = DataFrame(
+            [[1, 2, 3], [4, 5, 6]], columns=[["A", "A", "A"], ["a", "b", "c"]]
+        )
+        expected = DataFrame(
+            [[1, 2, 3], [4, 5, 6]],
+            columns=MultiIndex.from_tuples([("A", "a"), ("A", "b"), ("A", "c")]),
+        )
+        tm.assert_frame_equal(result, expected)

--- a/pandas/tests/frame/test_stack_unstack.py
+++ b/pandas/tests/frame/test_stack_unstack.py
@@ -1999,6 +1999,39 @@ Thu,Lunch,Yes,51.51,17"""
         )
         tm.assert_frame_equal(result, expected)
 
+    def test_multi_level_stack_categorical(self):
+        # GH 15239
+        midx = MultiIndex.from_arrays(
+            [
+                ["A"] * 2 + ["B"] * 2,
+                pd.Categorical(list("abab")),
+                pd.Categorical(list("ccdd")),
+            ]
+        )
+        df = DataFrame(np.arange(8).reshape(2, 4), columns=midx)
+        result = df.stack([1, 2])
+        expected = DataFrame(
+            [
+                [0, np.nan],
+                [np.nan, 2],
+                [1, np.nan],
+                [np.nan, 3],
+                [4, np.nan],
+                [np.nan, 6],
+                [5, np.nan],
+                [np.nan, 7],
+            ],
+            columns=["A", "B"],
+            index=MultiIndex.from_arrays(
+                [
+                    [0] * 4 + [1] * 4,
+                    pd.Categorical(list("aabbaabb")),
+                    pd.Categorical(list("cdcdcdcd")),
+                ]
+            ),
+        )
+        tm.assert_frame_equal(result, expected)
+
     def test_stack_nan_level(self):
         # GH 9406
         df_nan = DataFrame(

--- a/pandas/tests/groupby/test_groupby.py
+++ b/pandas/tests/groupby/test_groupby.py
@@ -2265,3 +2265,11 @@ def test_groupby_mean_duplicate_index(rand_series_with_duplicate_datetimeindex):
     result = dups.groupby(level=0).mean()
     expected = dups.groupby(dups.index).mean()
     tm.assert_series_equal(result, expected)
+
+
+def test_groupby_all_nan_groups_drop():
+    # GH 15036
+    s = Series([1, 2, 3], [np.nan, np.nan, np.nan])
+    result = s.groupby(s.index).sum()
+    expected = Series([], index=Index([], dtype=np.float64), dtype=np.int64)
+    tm.assert_series_equal(result, expected)

--- a/pandas/tests/groupby/test_groupby.py
+++ b/pandas/tests/groupby/test_groupby.py
@@ -2273,3 +2273,12 @@ def test_groupby_all_nan_groups_drop():
     result = s.groupby(s.index).sum()
     expected = Series([], index=Index([], dtype=np.float64), dtype=np.int64)
     tm.assert_series_equal(result, expected)
+
+
+def test_groupby_empty_multi_column():
+    # GH 15106
+    result = DataFrame(data=[], columns=["A", "B", "C"]).groupby(["A", "B"]).sum()
+    expected = DataFrame(
+        [], columns=["C"], index=MultiIndex([[], []], [[], []], names=["A", "B"])
+    )
+    tm.assert_frame_equal(result, expected)

--- a/pandas/tests/io/json/test_pandas.py
+++ b/pandas/tests/io/json/test_pandas.py
@@ -1750,3 +1750,23 @@ DataFrame\\.index values are different \\(100\\.0 %\\)
         result = read_json("[true, true, false]", typ="series")
         expected = Series([True, True, False])
         tm.assert_series_equal(result, expected)
+
+    def test_to_json_multiindex_escape(self):
+        # GH 15273
+        df = DataFrame(
+            True,
+            index=pd.date_range("2017-01-20", "2017-01-23"),
+            columns=["foo", "bar"],
+        ).stack()
+        result = df.to_json()
+        expected = (
+            "{\"(Timestamp('2017-01-20 00:00:00'), 'foo')\":true,"
+            "\"(Timestamp('2017-01-20 00:00:00'), 'bar')\":true,"
+            "\"(Timestamp('2017-01-21 00:00:00'), 'foo')\":true,"
+            "\"(Timestamp('2017-01-21 00:00:00'), 'bar')\":true,"
+            "\"(Timestamp('2017-01-22 00:00:00'), 'foo')\":true,"
+            "\"(Timestamp('2017-01-22 00:00:00'), 'bar')\":true,"
+            "\"(Timestamp('2017-01-23 00:00:00'), 'foo')\":true,"
+            "\"(Timestamp('2017-01-23 00:00:00'), 'bar')\":true}"
+        )
+        assert result == expected

